### DISCUSSION TITLE
Custom data type with None

### DISCUSSION THREAD

**Discussion by paultatulea** [pydantic/pydantic#3805] [Question] [answered]
Hello,

I want to create a custom data type that will coerce a None value to an empty string.

I am able to do this with a validator on a class with `pre=True` but I don't want to specify the validator on every class. Instead I want to use a custom data type which I believe is more clear.

In this example, `x` will not raise an error and be coerced to an empty string, but `y` will raise an error that None is not allowed.

```python
from typing import Optional

from pydantic.dataclasses import dataclass
from pydantic import validator
from pydantic.validators import str_validator


def none_to_empty_str(v: Optional[str]) -> str:
    if v is None:
        return ""
    return v


class NoneToEmptyStr(str):
    @classmethod
    def __get_validators__(cls):
        yield none_to_empty_str
        yield str_validator


@dataclass
class M:
    x: str
    y: NoneToEmptyStr

    _none_to_empty = validator("x", pre=True)(none_to_empty_str)


m = M(x=None, y=None)
```
```
pydantic.error_wrappers.ValidationError: 1 validation error for M
y
  none is not an allowed value (type=type_error.none.not_allowed)
```

I think the issue is that pydantic checks if None is allowed before running validators. Is there any way to achieve my custom data type?

**Comment by dbrattli** [comment]
I basically have the same issue with a custom [Option](https://github.com/cognitedata/Expression/blob/6181b7bd3379ed2275ab0d45be885e2336178fd3/expression/core/option.py#L53) type that should be erased to value or null on serialization. Parsing gives me `none is not an allowed value` since as you say the check seems to be done before the validator. Hope that someone can chime in on how to deal with this problem.

**Comment by bdoms** [comment, marked as answer]
To allow `None` the field should be marked as `Optional`. Then the problem becomes that Pydantic doesn't run validators by default when values aren't set, including when it's `None`. So you have to make sure your validator also has the `always` flag set - see https://pydantic-docs.helpmanual.io/usage/validators/#validate-always

Here's an example of how I got it working, simply by adding a validator to my class:

```python
from pydantic import BaseMode, validator

class Example(BaseModel):
    fool: Optional[str]

    @validator('foo', pre=True, always=True)
    def noneToEmpty(cls, v):
        return v or ''
```

**Reply by paultatulea** [reply]
This works great! I found that `Optional` is not required and it will work. Might be more clear in this case as the purpose is to always have a string for the field.
